e2e/e2e_test.go

```diff
@@ -87,17 +108,202 @@ var _ = Describe("Whereabouts functionality", func() {
+					dt = time.Now()
+					fmt.Printf("iteration %d of %d, %s\n", i, envVars.numThrashIter, dt.String())
+
+					// Set to 0 replicas
+					replicaCount = int32(0)
+					replicaSet, err = clientInfo.updateReplicaSet(replicaSetObject(
+						replicaCount,
+						podTierLabel(rsName),
+						podNetworkSelectionElements(testNetworkName),
+					))
+					Expect(err).NotTo(HaveOccurred())
+
+					Expect(WaitForReplicaSetSteadyState(clientInfo.Client, testNamespace, label, replicaSet, rsSteadyTimeout)).To(Succeed())
+
+					// Check for 0 IP pool allocations
+					ipPools, err := k8sClient.ListIPPools(context.Background())
+					Expect(err).To(BeNil())
+					Expect(ipPools[0].Allocations()).To(BeEmpty()) // get ipPool by name rather than by hoping index is right - TODO
+
+					// Increase to desired number of replicas for test
+					allPods, err := clientInfo.Client.CoreV1().Pods("").List(context.TODO(), metav1.ListOptions{})
+					Expect(err).NotTo(HaveOccurred())
+					maxReplicas := (envVars.numComputeNodes*maxPodsPerNode - (len(allPods.Items))) * envVars.fillPercentCapacity / 100
+					replicaCount = int32(maxReplicas)
+					replicaSet, err = clientInfo.updateReplicaSet(replicaSetObject(
+						replicaCount,
+						podTierLabel(rsName),
+						podNetworkSelectionElements(testNetworkName),
+					))
+					Expect(err).NotTo(HaveOccurred())
+
+					Expect(WaitForReplicaSetSteadyState(clientInfo.Client, testNamespace, label, replicaSet, rsSteadyTimeout)).To(Succeed())
+
+					Expect(iPPoolConsistency(envVars.kubeconfigPath, k8sClient, clientInfo.Client)).To(Succeed())
+				}
+			})
+		})
 	})
 })
 
-func clusterConfig() (*rest.Config, error) {
+func iPPoolConsistency(kubeconfigPath string, k8sClient *kubeClient.Client, cs *kubernetes.Clientset) error {
+	By("checking if there are any stale IPs in IP pool or any IPs in IP pool that are not seen attached to a pod")
+
+	By("Forcing reconciliation of the cluster...")
+	if err := runIpReconciler(kubeconfigPath); err != nil {
+		return err
+	}
+
+	By("Comparing pod ip assignments to ippool reservations")
+	// TODO - stop this ipPools[0] stuff and get it by name if there's a way to do it. low priority though
+	ipPools, err := k8sClient.ListIPPools(context.Background())
+	if err != nil {
+		return err
+	}
+
+	ipPoolAllocations := ipPools[0].Allocations() // todo, get this more elegantly than ipPools[0]
+	podList, err := ListPods(cs, testNamespace, wbLabelEqual)
+	if err != nil {
+		return err
+	}
+
+	found := true
+	podIP := ""
+	reservedIP := ""
+
+	for _, allocation := range ipPoolAllocations {
+		found = false
+		for _, pod := range podList.Items {
+			podIP, err = secondaryIfaceIPValue(&pod)
+			if err != nil {
+				return err
+			}
+			reservedIP = allocation.IP.String()
+
+			if reservedIP == podIP {
+				found = true
+				break
+			}
+		}
+
+		if !found {
+			return fmt.Errorf("Possible stale IP Pool: failed to find pod for IP Pool IP %s\n", allocation.IP)
+		}
```

Comment:
you could export this into an helper func.

We could eventually think of a way to make this perform better than O(n^2), but right now, that's a premature optimization.

Reply:
I would be interested in hearing how this could perform better than O(n^2) :thinking: